File: agent/src/utils/circuit-breaker.ts
export enum CircuitState {
  Closed = 'closed',
  Open = 'open',
  HalfOpen = 'half_open',
}

export interface CircuitStats {
  state: CircuitState;
  failureCount: number;
  successCount: number;
  lastFailureTime: number | null;
  lastFailureMessage: string | null;
  openedAt: number | null;
  totalCalls: number;
  totalFailures: number;
  totalSuccesses: number;
}

export class CircuitOpenError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'CircuitOpenError';
  }
}

export class CircuitBreaker {
  private state: CircuitState = CircuitState.Closed;
  private failureCount = 0;
  private successCount = 0;
  private lastFailureTime: number | null = null;
  private lastFailureMessage: string | null = null;
  private openedAt: number | null = null;
  private halfOpenCalls = 0;
  private totalCalls = 0;
  private totalFailures = 0;
  private totalSuccesses = 0;

  constructor(
    private readonly name: string,
    private readonly failureThreshold: number = 5,
    private readonly recoveryTimeout: number = 30000,
    private readonly halfOpenMaxCalls: number = 3
  ) {}

  async call<T>(fn: () => Promise<T>): Promise<T> {
    this.maybeTransition();

    if (this.state === CircuitState.Open) {
      throw new CircuitOpenError(
        `Circuit '${this.name}' is open. Last failure: ${this.lastFailureMessage || 'unknown'}`
      );
    }

    if (this.state === CircuitState.HalfOpen) {
      if (this.halfOpenCalls >= this.halfOpenMaxCalls) {
        throw new CircuitOpenError(
          `Circuit '${this.name}' is in half-open state with max calls reached`
        );
      }
      this.halfOpenCalls++;
    }

    this.totalCalls++;

    try {
      const result = await fn();
      this.onSuccess();
      return result;
    } catch (error) {
      this.onFailure(error instanceof Error ? error.message : String(error));
      throw error;
    }
  }

  private maybeTransition(): void {
    if (this.state === CircuitState.Open && this.openedAt !== null) {
      if (Date.now() - this.openedAt >= this.recoveryTimeout) {
        this.state = CircuitState.HalfOpen;
        this.halfOpenCalls = 0;
      }
    }
  }

  private onSuccess(): void {
    this.successCount++;
    this.totalSuccesses++;

    if (this.state === CircuitState.HalfOpen) {
      if (this.successCount >= this.halfOpenMaxCalls) {
        this.reset();
      }
    }
  }

  private onFailure(message: string): void {
    this.failureCount++;
    this.totalFailures++;
    this.lastFailureTime = Date.now();
    this.lastFailureMessage = message;

    if (this.state === CircuitState.HalfOpen) {
      this.trip();
    } else if (this.failureCount >= this.failureThreshold) {
      this.trip();
    }
  }

  private trip(): void {
    this.state = CircuitState.Open;
    this.openedAt = Date.now();
  }

  private reset(): void {
    this.state = CircuitState.Closed;
    this.failureCount = 0;
    this.successCount = 0;
    this.openedAt = null;
  }

  forceOpen(): void {
    this.trip();
  }

  forceClose(): void {
    this.reset();
  }

  getStats(): CircuitStats {
    return {
      state: this.state,
      failureCount: this.failureCount,
      successCount: this.successCount,
      lastFailureTime: this.lastFailureTime,
      lastFailureMessage: this.lastFailureMessage,
      openedAt: this.openedAt,
      totalCalls: this.totalCalls,
      totalFailures: this.totalFailures,
      totalSuccesses: this.totalSuccesses,
    };
  }

  isOpen(): boolean {
    this.maybeTransition();
    return this.state === CircuitState.Open;
  }

  isClosed(): boolean {
    this.maybeTransition();
    return this.state === CircuitState.Closed;
  }
}

export class CircuitBreakerRegistry {
  private breakers: Map<string, CircuitBreaker> = new Map();

  getOrCreate(
    name: string,
    failureThreshold: number = 5,
    recoveryTimeout: number = 30000
  ): CircuitBreaker {
    let breaker = this.breakers.get(name);
    if (!breaker) {
      breaker = new CircuitBreaker(name, failureThreshold, recoveryTimeout);
      this.breakers.set(name, breaker);
    }
    return breaker;
  }

  get(name: string): CircuitBreaker | undefined {
    return this.breakers.get(name);
  }

  getAllStats(): Record<string, CircuitStats> {
    const stats: Record<string, CircuitStats> = {};
    for (const [name, breaker] of this.breakers) {
      stats[name] = breaker.getStats();
    }
    return stats;
  }

  getHealth(): { healthy: boolean; totalCircuits: number; openCircuits: string[] } {
    const stats = this.getAllStats();
    const openCircuits = Object.entries(stats)
      .filter(([, s]) => s.state === CircuitState.Open)
      .map(([name]) => name);

    return {
      healthy: openCircuits.length === 0,
      totalCircuits: Object.keys(stats).length,
      openCircuits,
    };
  }
}

export const circuitRegistry = new CircuitBreakerRegistry();

export function withCircuitBreaker<T>(
  name: string,
  fn: () => Promise<T>,
  options?: { failureThreshold?: number; recoveryTimeout?: number }
): Promise<T> {
  const breaker = circuitRegistry.getOrCreate(
    name,
    options?.failureThreshold,
    options?.recoveryTimeout
  );
  return breaker.call(fn);
}

File: core/src/scholardevclaw/utils/circuit_breaker.py
from __future__ import annotations

import time
import threading
from dataclasses import dataclass, field
from datetime import datetime
from enum import Enum
from typing import Any, Callable, Generic, TypeVar

T = TypeVar("T")


class CircuitState(Enum):
    CLOSED = "closed"
    OPEN = "open"
    HALF_OPEN = "half_open"


@dataclass
class CircuitStats:
    state: CircuitState
    failure_count: int
    success_count: int
    last_failure_time: float | None
    last_failure_message: str | None
    opened_at: float | None
    total_calls: int
    total_failures: int
    total_successes: int


class CircuitBreaker(Generic[T]):
    """Circuit breaker for external service calls."""

    def __init__(
        self,
        name: str,
        failure_threshold: int = 5,
        recovery_timeout: float = 30.0,
        half_open_max_calls: int = 3,
    ):
        self.name = name
        self.failure_threshold = failure_threshold
        self.recovery_timeout = recovery_timeout
        self.half_open_max_calls = half_open_max_calls

        self._state = CircuitState.CLOSED
        self._failure_count = 0
        self._success_count = 0
        self._last_failure_time: float | None = None
        self._last_failure_message: str | None = None
        self._opened_at: float | None = None
        self._half_open_calls = 0
        self._lock = threading.RLock()

        self._total_calls = 0
        self._total_failures = 0
        self._total_successes = 0

    @property
    def state(self) -> CircuitState:
        with self._lock:
            self._maybe_transition()
            return self._state

    def _maybe_transition(self) -> None:
        if self._state == CircuitState.OPEN:
            if self._opened_at and time.time() - self._opened_at >= self.recovery_timeout:
                self._state = CircuitState.HALF_OPEN
                self._half_open_calls = 0

    def call(self, func: Callable[[], T]) -> T:
        with self._lock:
            self._maybe_transition()

            if self._state == CircuitState.OPEN:
                raise CircuitOpenError(
                    f"Circuit '{self.name}' is open. "
                    f"Last failure: {self._last_failure_message or 'unknown'}"
                )

            if self._state == CircuitState.HALF_OPEN:
                if self._half_open_calls >= self.half_open_max_calls:
                    raise CircuitOpenError(
                        f"Circuit '{self.name}' is in half-open state with max calls reached"
                    )
                self._half_open_calls += 1

        self._total_calls += 1

        try:
            result = func()
            self._on_success()
            return result
        except Exception as e:
            self._on_failure(e)
            raise

    def _on_success(self) -> None:
        with self._lock:
            self._success_count += 1
            self._total_successes += 1

            if self._state == CircuitState.HALF_OPEN:
                if self._success_count >= self.half_open_max_calls:
                    self._reset()

    def _on_failure(self, error: Exception) -> None:
        with self._lock:
            self._failure_count += 1
            self._total_failures += 1
            self._last_failure_time = time.time()
            self._last_failure_message = str(error)

            if self._state == CircuitState.HALF_OPEN:
                self._trip()
            elif self._failure_count >= self.failure_threshold:
                self._trip()

    def _trip(self) -> None:
        self._state = CircuitState.OPEN
        self._opened_at = time.time()

    def _reset(self) -> None:
        self._state = CircuitState.CLOSED
        self._failure_count = 0
        self._success_count = 0
        self._opened_at = None

    def force_open(self) -> None:
        with self._lock:
            self._trip()

    def force_close(self) -> None:
        with self._lock:
            self._reset()

    def get_stats(self) -> CircuitStats:
        with self._lock:
            return CircuitStats(
                state=self._state,
                failure_count=self._failure_count,
                success_count=self._success_count,
                last_failure_time=self._last_failure_time,
                last_failure_message=self._last_failure_message,
                opened_at=self._opened_at,
                total_calls=self._total_calls,
                total_failures=self._total_failures,
                total_successes=self._total_successes,
            )


class CircuitOpenError(Exception):
    """Raised when circuit breaker is open."""

    pass


class CircuitBreakerRegistry:
    """Registry for managing multiple circuit breakers."""

    def __init__(self):
        self._breakers: dict[str, CircuitBreaker] = {}
        self._lock = threading.RLock()

    def get_or_create(
        self,
        name: str,
        failure_threshold: int = 5,
        recovery_timeout: float = 30.0,
    ) -> CircuitBreaker:
        with self._lock:
            if name not in self._breakers:
                self._breakers[name] = CircuitBreaker(
                    name=name,
                    failure_threshold=failure_threshold,
                    recovery_timeout=recovery_timeout,
                )
            return self._breakers[name]

    def get(self, name: str) -> CircuitBreaker | None:
        with self._lock:
            return self._breakers.get(name)

    def get_all_stats(self) -> dict[str, CircuitStats]:
        with self._lock:
            return {name: breaker.get_stats() for name, breaker in self._breakers.items()}

    def get_health(self) -> dict[str, Any]:
        stats = self.get_all_stats()
        open_circuits = [name for name, s in stats.items() if s.state == CircuitState.OPEN]

        return {
            "healthy": len(open_circuits) == 0,
            "total_circuits": len(stats),
            "open_circuits": open_circuits,
            "stats": {
                name: {
                    "state": s.state.value,
                    "failure_count": s.failure_count,
                    "total_calls": s.total_calls,
                }
                for name, s in stats.items()
            },
        }


circuit_registry = CircuitBreakerRegistry()


def with_circuit_breaker(
    name: str,
    failure_threshold: int = 5,
    recovery_timeout: float = 30.0,
):
    """Decorator to wrap function with circuit breaker."""

    def decorator(func: Callable[[], T]) -> Callable[[], T]:
        breaker = circuit_registry.get_or_create(
            name,
            failure_threshold=failure_threshold,
            recovery_timeout=recovery_timeout,
        )

        def wrapper(*args, **kwargs):
            return breaker.call(lambda: func(*args, **kwargs))

        return wrapper

    return decorator
